File: ccf_experiments.py
"""
CCF Experiments: Benchmarking CCF algorithms on synthetic graphs
================================================================
Generates synthetic graphs of increasing size and topology, runs both
CCF-Iterate variants, and records iterations and runtime.
"""

import time
import random
import csv
from pyspark import SparkContext, SparkConf


# =============================================================================
# Graph Generators
# =============================================================================

def generate_chain_graph(n):
    """Chain: 0-1-2-...-n-1. Diameter = n-1 (worst case for iterations)."""
    return [(str(i), str(i + 1)) for i in range(n - 1)]


def generate_random_graph(n_nodes, n_edges, seed=42):
    """Erdos-Renyi style random graph with fixed number of edges."""
    rng = random.Random(seed)
    edges = set()
    while len(edges) < n_edges:
        a = rng.randint(0, n_nodes - 1)
        b = rng.randint(0, n_nodes - 1)
        if a != b:
            edge = (str(min(a, b)), str(max(a, b)))
            edges.add(edge)
    return list(edges)


def generate_cluster_graph(n_clusters, nodes_per_cluster, inter_edges=0, seed=42):
    """Multiple dense clusters with optional inter-cluster edges."""
    rng = random.Random(seed)
    edges = []
    # Intra-cluster: connect each node to ~3 neighbors within cluster
    for c in range(n_clusters):
        base = c * nodes_per_cluster
        for i in range(nodes_per_cluster - 1):
            edges.append((str(base + i), str(base + i + 1)))
            # Add a few random intra-cluster edges for density
            if i + 2 < nodes_per_cluster:
                edges.append((str(base + i), str(base + i + 2)))
    # Inter-cluster edges
    for _ in range(inter_edges):
        c1, c2 = rng.sample(range(n_clusters), 2)
        n1 = c1 * nodes_per_cluster + rng.randint(0, nodes_per_cluster - 1)
        n2 = c2 * nodes_per_cluster + rng.randint(0, nodes_per_cluster - 1)
        edges.append((str(n1), str(n2)))
    return edges


# =============================================================================
# CCF Implementations (inlined for experiment)
# =============================================================================

def ccf_iterate_basic(pairs_rdd):
    mapped = pairs_rdd.flatMap(lambda pair: [(pair[0], pair[1]), (pair[1], pair[0])])
    grouped = mapped.groupByKey()

    def reduce_fn(kv):
        key, values_iter = kv
        value_list = list(values_iter)
        min_val = key
        for v in value_list:
            if v < min_val:
                min_val = v
        results = []
        new_pairs = 0
        if min_val < key:
            results.append((key, min_val))
            for v in value_list:
                if v != min_val:
                    new_pairs += 1
                    results.append((v, min_val))
        return results, new_pairs

    reduced = grouped.map(reduce_fn).cache()
    output_rdd = reduced.flatMap(lambda x: x[0])
    new_pair_count = reduced.map(lambda x: x[1]).reduce(lambda a, b: a + b)
    reduced.unpersist()
    return output_rdd, new_pair_count


def ccf_iterate_secondary_sort(pairs_rdd):
    mapped = pairs_rdd.flatMap(lambda pair: [(pair[0], pair[1]), (pair[1], pair[0])])
    grouped = mapped.groupByKey().mapValues(lambda vals: sorted(vals))

    def reduce_fn(kv):
        key, sorted_values = kv
        results = []
        new_pairs = 0
        if not sorted_values:
            return results, new_pairs
        min_value = sorted_values[0]
        if min_value < key:
            results.append((key, min_value))
            for v in sorted_values[1:]:
                new_pairs += 1
                results.append((v, min_value))
        return results, new_pairs

    reduced = grouped.map(reduce_fn).cache()
    output_rdd = reduced.flatMap(lambda x: x[0])
    new_pair_count = reduced.map(lambda x: x[1]).reduce(lambda a, b: a + b)
    reduced.unpersist()
    return output_rdd, new_pair_count


def ccf_dedup(pairs_rdd):
    return pairs_rdd.distinct()


def run_ccf(sc, edges, iterate_fn, max_iterations=100):
    """Run full CCF pipeline, return (iterations, runtime_seconds, num_components)."""
    start = time.time()
    pairs_rdd = sc.parallelize(edges)
    iteration = 0

    while iteration < max_iterations:
        iteration += 1
        output_rdd, new_pair_count = iterate_fn(pairs_rdd)
        pairs_rdd = ccf_dedup(output_rdd)
        pairs_rdd.cache()
        pairs_rdd.count()  # force evaluation

        if new_pair_count == 0:
            break

    elapsed = time.time() - start

    # Count components
    num_components = pairs_rdd.map(lambda x: x[1]).distinct().count()

    return iteration, elapsed, num_components


# =============================================================================
# Experiment Runner
# =============================================================================

def main():
    conf = SparkConf().setAppName("CCF-Experiments").setMaster("local[*]")
    conf.set("spark.ui.showConsoleProgress", "false")
    sc = SparkContext(conf=conf)
    sc.setLogLevel("ERROR")

    results = []

    # ---- Experiment 1: Increasing graph size (random graphs) ----
    print("=" * 70)
    print("EXPERIMENT 1: Random graphs of increasing size")
    print("=" * 70)
    print(f"{'Nodes':>8} {'Edges':>8} {'Algorithm':>15} {'Iters':>6} {'Time(s)':>10} {'Components':>12}")
    print("-" * 70)

    random_configs = [
        (50,    100),
        (100,   300),
        (500,   1500),
        (1000,  3000),
        (2000,  6000),
        (5000,  15000),
    ]

    for n_nodes, n_edges in random_configs:
        edges = generate_random_graph(n_nodes, n_edges)

        for name, fn in [("Basic", ccf_iterate_basic), ("SecondarySort", ccf_iterate_secondary_sort)]:
            iters, elapsed, n_comp = run_ccf(sc, edges, fn)
            print(f"{n_nodes:>8} {n_edges:>8} {name:>15} {iters:>6} {elapsed:>10.3f} {n_comp:>12}")
            results.append({
                "experiment": "random_graph",
                "nodes": n_nodes,
                "edges": n_edges,
                "algorithm": name,
                "iterations": iters,
                "runtime_sec": round(elapsed, 3),
                "components": n_comp,
            })

    # ---- Experiment 2: Chain graphs (worst case diameter) ----
    print()
    print("=" * 70)
    print("EXPERIMENT 2: Chain graphs (worst-case diameter)")
    print("=" * 70)
    print(f"{'Nodes':>8} {'Edges':>8} {'Algorithm':>15} {'Iters':>6} {'Time(s)':>10} {'Components':>12}")
    print("-" * 70)

    chain_sizes = [10, 50, 100, 200, 500]

    for n in chain_sizes:
        edges = generate_chain_graph(n)

        for name, fn in [("Basic", ccf_iterate_basic), ("SecondarySort", ccf_iterate_secondary_sort)]:
            iters, elapsed, n_comp = run_ccf(sc, edges, fn)
            print(f"{n:>8} {n-1:>8} {name:>15} {iters:>6} {elapsed:>10.3f} {n_comp:>12}")
            results.append({
                "experiment": "chain_graph",
                "nodes": n,
                "edges": n - 1,
                "algorithm": name,
                "iterations": iters,
                "runtime_sec": round(elapsed, 3),
                "components": n_comp,
            })

    # ---- Experiment 3: Cluster graphs ----
    print()
    print("=" * 70)
    print("EXPERIMENT 3: Cluster graphs (multiple components)")
    print("=" * 70)
    print(f"{'Clusters':>9} {'Nodes/C':>8} {'Inter':>6} {'Algorithm':>15} {'Iters':>6} {'Time(s)':>10} {'Components':>12}")
    print("-" * 70)

    cluster_configs = [
        (5,   20,  0),    # 5 isolated clusters
        (5,   20,  4),    # 5 clusters, some connected
        (10,  50,  0),    # 10 isolated clusters
        (10,  50,  9),    # 10 clusters, all connected
        (20,  50,  0),    # 20 isolated clusters
        (20,  50,  19),   # 20 clusters, all connected
    ]

    for n_clusters, npc, inter in cluster_configs:
        edges = generate_cluster_graph(n_clusters, npc, inter)
        total_nodes = n_clusters * npc

        for name, fn in [("Basic", ccf_iterate_basic), ("SecondarySort", ccf_iterate_secondary_sort)]:
            iters, elapsed, n_comp = run_ccf(sc, edges, fn)
            print(f"{n_clusters:>9} {npc:>8} {inter:>6} {name:>15} {iters:>6} {elapsed:>10.3f} {n_comp:>12}")
            results.append({
                "experiment": "cluster_graph",
                "nodes": total_nodes,
                "edges": len(edges),
                "algorithm": name,
                "iterations": iters,
                "runtime_sec": round(elapsed, 3),
                "components": n_comp,
                "clusters": n_clusters,
                "inter_edges": inter,
            })

    # ---- Save results to CSV ----
    csv_path = "/Users/davidwfong/Code/Master/map-reduce-project/experiment_results.csv"
    fieldnames = ["experiment", "nodes", "edges", "algorithm", "iterations",
                  "runtime_sec", "components", "clusters", "inter_edges"]
    with open(csv_path, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames, extrasaction="ignore")
        writer.writeheader()
        for row in results:
            writer.writerow(row)

    print(f"\nResults saved to {csv_path}")

    sc.stop()


if __name__ == "__main__":
    main()

File: CCFExperiments.scala
/**
 * CCF Experiments: Benchmarking on synthetic graphs
 * ==================================================
 * Mirrors ccf_experiments.py — runs Basic and SecondarySort variants
 * on random, chain, and cluster graphs; records iterations, runtime, components.
 *
 * Run with:
 *   spark-submit --class CCFExperiments CCFExperiments.jar
 * or via spark-shell / spark-submit with --packages
 */

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.rdd.RDD
import org.apache.spark.util.LongAccumulator

import scala.collection.mutable
import scala.util.Random

object CCFExperiments {

  // ===========================================================================
  // Graph Generators (mirrors ccf_experiments.py)
  // ===========================================================================

  def generateChainGraph(n: Int): Seq[(String, String)] =
    (0 until n - 1).map(i => (i.toString, (i + 1).toString))

  def generateRandomGraph(nNodes: Int, nEdges: Int, seed: Int = 42): Seq[(String, String)] = {
    val rng = new Random(seed)
    val edges = mutable.Set[(String, String)]()
    while (edges.size < nEdges) {
      val a = rng.nextInt(nNodes)
      val b = rng.nextInt(nNodes)
      if (a != b) {
        val lo = math.min(a, b).toString
        val hi = math.max(a, b).toString
        edges.add((lo, hi))
      }
    }
    edges.toSeq
  }

  def generateClusterGraph(nClusters: Int, nodesPerCluster: Int,
                           interEdges: Int = 0, seed: Int = 42): Seq[(String, String)] = {
    val rng = new Random(seed)
    val edges = mutable.ListBuffer[(String, String)]()
    for (c <- 0 until nClusters) {
      val base = c * nodesPerCluster
      for (i <- 0 until nodesPerCluster - 1) {
        edges += ((( base + i).toString, (base + i + 1).toString))
        if (i + 2 < nodesPerCluster)
          edges += (((base + i).toString, (base + i + 2).toString))
      }
    }
    for (_ <- 0 until interEdges) {
      val cs = rng.shuffle((0 until nClusters).toList).take(2)
      val n1 = cs(0) * nodesPerCluster + rng.nextInt(nodesPerCluster)
      val n2 = cs(1) * nodesPerCluster + rng.nextInt(nodesPerCluster)
      edges += ((n1.toString, n2.toString))
    }
    edges.toSeq
  }

  // ===========================================================================
  // CCF Algorithm (from CCFConnectedComponents.scala)
  // ===========================================================================

  def ccfIterate(pairsRDD: RDD[(String, String)],
                 counter: LongAccumulator): RDD[(String, String)] = {
    val mapped = pairsRDD.flatMap { case (k, v) => Seq((k, v), (v, k)) }
    mapped.groupByKey().flatMap { case (key, valuesIter) =>
      val valueList = valuesIter.toList
      var min = key
      for (v <- valueList) if (v < min) min = v
      if (min < key) {
        val results = mutable.ListBuffer((key, min))
        for (v <- valueList) if (v != min) { counter.add(1); results += ((v, min)) }
        results.toList
      } else List.empty
    }
  }

  def ccfIterateSecondarySort(pairsRDD: RDD[(String, String)],
                              counter: LongAccumulator): RDD[(String, String)] = {
    val mapped = pairsRDD.flatMap { case (k, v) => Seq((k, v), (v, k)) }
    mapped.groupByKey().flatMap { case (key, valuesIter) =>
      val sorted = valuesIter.toList.sorted
      if (sorted.nonEmpty && sorted.head < key) {
        val minValue = sorted.head
        val results = mutable.ListBuffer((key, minValue))
        for (v <- sorted.tail) { counter.add(1); results += ((v, minValue)) }
        results.toList
      } else List.empty
    }
  }

  def ccfDedup(pairsRDD: RDD[(String, String)]): RDD[(String, String)] =
    pairsRDD.map(p => (p, null)).reduceByKey((a, _) => a).map(_._1)

  // ===========================================================================
  // Experiment Runner
  // ===========================================================================

  case class Result(experiment: String, nodes: Int, edges: Int, algorithm: String,
                    iterations: Int, runtimeSec: Double, components: Int,
                    clusters: Int = 0, interEdges: Int = 0)

  def runCCF(sc: SparkContext, edges: Seq[(String, String)],
             useSecondarySort: Boolean, maxIterations: Int = 100): (Int, Double, Int) = {
    val t0 = System.currentTimeMillis()
    var pairsRDD = sc.parallelize(edges)
    var iteration = 0
    var converged = false

    while (iteration < maxIterations && !converged) {
      iteration += 1
      val counter = sc.longAccumulator("NewPair")
      val output = if (useSecondarySort) ccfIterateSecondarySort(pairsRDD, counter)
                   else ccfIterate(pairsRDD, counter)
      val deduped = ccfDedup(output)
      deduped.cache()
      deduped.count()
      pairsRDD.unpersist()
      pairsRDD = deduped
      if (counter.value == 0) converged = true
    }

    val elapsed = (System.currentTimeMillis() - t0) / 1000.0
    val nComponents = pairsRDD.map(_._2).distinct().count().toInt
    (iteration, elapsed, nComponents)
  }

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("CCF-Experiments").setMaster("local[*]")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    val results = mutable.ListBuffer[Result]()

    // ---- Experiment 1: Random graphs ----
    println("=" * 70)
    println("EXPERIMENT 1: Random graphs of increasing size")
    println("=" * 70)
    println(f"${"Nodes"}%8s ${"Edges"}%8s ${"Algorithm"}%15s ${"Iters"}%6s ${"Time(s)"}%10s ${"Components"}%12s")
    println("-" * 70)

    val randomConfigs = Seq((50, 100), (100, 300), (500, 1500), (1000, 3000), (2000, 6000), (5000, 15000))
    for ((nNodes, nEdges) <- randomConfigs) {
      val edges = generateRandomGraph(nNodes, nEdges)
      for ((name, useSecSort) <- Seq(("Basic", false), ("SecondarySort", true))) {
        val (iters, elapsed, nComp) = runCCF(sc, edges, useSecSort)
        println(f"$nNodes%8d $nEdges%8d $name%15s $iters%6d $elapsed%10.3f $nComp%12d")
        results += Result("random_graph", nNodes, nEdges, name, iters, elapsed, nComp)
      }
    }

    // ---- Experiment 2: Chain graphs ----
    println()
    println("=" * 70)
    println("EXPERIMENT 2: Chain graphs (worst-case diameter)")
    println("=" * 70)
    println(f"${"Nodes"}%8s ${"Edges"}%8s ${"Algorithm"}%15s ${"Iters"}%6s ${"Time(s)"}%10s ${"Components"}%12s")
    println("-" * 70)

    val chainSizes = Seq(10, 50, 100, 200, 500)
    for (n <- chainSizes) {
      val edges = generateChainGraph(n)
      for ((name, useSecSort) <- Seq(("Basic", false), ("SecondarySort", true))) {
        val (iters, elapsed, nComp) = runCCF(sc, edges, useSecSort)
        println(f"$n%8d ${n - 1}%8d $name%15s $iters%6d $elapsed%10.3f $nComp%12d")
        results += Result("chain_graph", n, n - 1, name, iters, elapsed, nComp)
      }
    }

    // ---- Experiment 3: Cluster graphs ----
    println()
    println("=" * 70)
    println("EXPERIMENT 3: Cluster graphs (multiple components)")
    println("=" * 70)
    println(f"${"Clusters"}%9s ${"Nodes/C"}%8s ${"Inter"}%6s ${"Algorithm"}%15s ${"Iters"}%6s ${"Time(s)"}%10s ${"Components"}%12s")
    println("-" * 70)

    val clusterConfigs = Seq((5, 20, 0), (5, 20, 4), (10, 50, 0), (10, 50, 9), (20, 50, 0), (20, 50, 19))
    for ((nClusters, npc, inter) <- clusterConfigs) {
      val edges = generateClusterGraph(nClusters, npc, inter)
      for ((name, useSecSort) <- Seq(("Basic", false), ("SecondarySort", true))) {
        val (iters, elapsed, nComp) = runCCF(sc, edges, useSecSort)
        println(f"$nClusters%9d $npc%8d $inter%6d $name%15s $iters%6d $elapsed%10.3f $nComp%12d")
        results += Result("cluster_graph", nClusters * npc, edges.size, name, iters, elapsed, nComp, nClusters, inter)
      }
    }

    // ---- Write CSV ----
    val csvPath = "/Users/davidwfong/Code/Master/map-reduce-project/experiment_results_scala.csv"
    val pw = new java.io.PrintWriter(csvPath)
    pw.println("experiment,nodes,edges,algorithm,iterations,runtime_sec,components,clusters,inter_edges")
    for (r <- results) {
      pw.println(s"${r.experiment},${r.nodes},${r.edges},${r.algorithm},${r.iterations},${r.runtimeSec},${r.components},${r.clusters},${r.interEdges}")
    }
    pw.close()
    println(s"\nResults saved to $csvPath")

    sc.stop()
  }
}
